File: services/route/varuna_route/profiles.py
"""Vehicle profiles: the depth at which each one stops (CLAUDE.md 11.8, Appendix A).

The thresholds are the depth ramp of CLAUDE.md 6.2, which is the same set of numbers the map
colours by and the alert levels raise at. That is deliberate: an operator who has learnt that
orange means "cars stop" should not have to learn a second scale to read a route.

Risk tolerance is the probability a driver will accept of meeting more than their threshold.
An ambulance's 0.2 is not caution, it is the opposite - it means the ambulance refuses a street
that has even a one-in-five chance of stopping it, because an ambulance stopped in water is a
second emergency. A car's 0.5 is an ordinary driver's coin flip.
"""

from __future__ import annotations

from dataclasses import dataclass

__all__ = ["PROFILES", "Profile", "profile"]


@dataclass(frozen=True, slots=True)
class Profile:
    """One vehicle's rules for water."""

    key: str
    label: str
    depth_cm: float
    """Depth at which this vehicle can no longer pass."""

    risk_tolerance: float
    """Probability of exceeding ``depth_cm`` that the route will accept on an edge."""

    speed_scale: float
    """Multiplier on the road's free-flow speed: an ambulance is quicker than a bus."""

    hazard_rule: bool = False
    """Pedestrians also fail on the ``h * v >= 0.5 m^2/s`` hazard product (Appendix A)."""


PROFILES: dict[str, Profile] = {
    # 15 cm: `--depth-2` begins, the ramp's "two-wheelers impassable".
    "two_wheeler": Profile("two_wheeler", "Two-wheeler", 15.0, 0.5, 1.0),
    # 30 cm: `--depth-3`, "cars impassable".
    "car": Profile("car", "Car", 30.0, 0.5, 1.0),
    # 45 cm: `--depth-4`, "buses and trucks impassable".
    "bus": Profile("bus", "Bus", 45.0, 0.4, 0.8),
    "truck": Profile("truck", "Truck", 45.0, 0.4, 0.8),
    # 60 cm: `--depth-5`, "rescue vehicles only" - and this is the rescue vehicle.
    "ambulance": Profile("ambulance", "Ambulance", 60.0, 0.2, 1.15),
    "fire_tender": Profile("fire_tender", "Fire tender", 60.0, 0.2, 1.0),
    # A person is stopped by 30 cm, and by less when it moves (the hazard product below).
    "pedestrian": Profile("pedestrian", "Pedestrian", 30.0, 0.3, 0.06, hazard_rule=True),
}
"""Every profile CLAUDE.md 7.4 puts in the picker, keyed by the value the API takes."""


def profile(key: str) -> Profile:
    """Look one up, raising :class:`KeyError` with the valid set when it is not one."""
    try:
        return PROFILES[key]
    except KeyError:
        valid = ", ".join(sorted(PROFILES))
        msg = f"Unknown vehicle profile {key!r}. Valid profiles: {valid}."
        raise KeyError(msg) from None

File: services/route-rs/src/profiles.rs
//! Vehicle profiles (`varuna_route.profiles`).
//!
//! The table the service routes with is the one in the graph export, written from the Python
//! `PROFILES` dict, so a threshold changed in Python reaches Rust on the next export rather than
//! drifting. [`defaults`] is a copy for unit tests only, and a test pins it to the export.

use serde::Deserialize;

#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct Profile {
    pub key: String,
    pub label: String,
    /// Depth at which this vehicle can no longer pass.
    pub depth_cm: f64,
    /// Probability of exceeding `depth_cm` the route will accept on an edge.
    pub risk_tolerance: f64,
    /// Multiplier on free-flow speed.
    pub speed_scale: f64,
    #[serde(default)]
    pub hazard_rule: bool,
}

/// The Python table as of 2026-09-19, for tests that run without an export.
pub fn defaults() -> Vec<Profile> {
    let p = |key: &str, label: &str, depth: f64, tol: f64, speed: f64, hazard: bool| Profile {
        key: key.into(),
        label: label.into(),
        depth_cm: depth,
        risk_tolerance: tol,
        speed_scale: speed,
        hazard_rule: hazard,
    };
    vec![
        p("two_wheeler", "Two-wheeler", 15.0, 0.5, 1.0, false),
        p("car", "Car", 30.0, 0.5, 1.0, false),
        p("bus", "Bus", 45.0, 0.4, 0.8, false),
        p("truck", "Truck", 45.0, 0.4, 0.8, false),
        p("ambulance", "Ambulance", 60.0, 0.2, 1.15, false),
        p("fire_tender", "Fire tender", 60.0, 0.2, 1.0, false),
        p("pedestrian", "Pedestrian", 30.0, 0.3, 0.06, true),
    ]
}

/// The valid keys, sorted, as the Python error message lists them.
pub fn valid_keys(profiles: &[Profile]) -> String {
    let mut keys: Vec<&str> = profiles.iter().map(|p| p.key.as_str()).collect();
    keys.sort_unstable();
    keys.join(", ")
}
